Refactor stack handling to always use TracingContext to populate real stack on exception

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106403
* __->__ #106277

The basic gist of the PR is simple, but it's accompanied with some careful modifications and unit tests to make sure I got it right. Check inline comments for more details.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

ezyang committed:
Refactor stack handling to always use TracingContext to populate real stack on exception

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

albanD reviewed (approved):
Sounds good

ezyang committed:
Update on "Refactor stack handling to always use TracingContext to populate real stack on exception"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/_dynamo/exc.py:
```diff
@@ -67,7 +69,7 @@ def __init__(self, backend_fn, inner_exception):
     def __init__(self, msg):
         super().__init__(msg)
-        self.real_stack = []
+        self.real_stack = torch._guards.TracingContext.extract_stack()
```
Instead of adding stack entries as we walk up the exception frames, we always have an accurate user stack at any given point in time, so we can just use it directly. Making exceptions make use of this facility is not strictly necessary, but stylistically coherent.

ezyang commented on torch/_dynamo/exc.py:
```diff
@@ -162,12 +164,11 @@ def __repr__(self) -> str:
-        and len(exc.real_stack) > 0
-        and not (config.verbose and config.suppress_errors)
-    ):
-        msg += f"\nfrom user code:\n {''.join(traceback.format_list(list(reversed(get_real_stack(exc)[0:2]))))}"
```
I don't really know why the real stack used to be truncated to two entries; I opt for printing the entire stack because... why would you NOT report the full stack 🤔

ezyang replied:
This case is tested for in test_unsupported_real_stack among others.

ezyang commented on torch/_dynamo/exc.py:
```diff
@@ -208,9 +209,20 @@ def augment_exc_message(exc, msg="\n", export=False):
+    if frame is not None:
+        stack_above_dynamo = filter_stack(extract_stack(frame))
+
+    return cast(StackSummary, stack_above_dynamo + real_stack)
```
Allowing this function to return None simplifies a bunch of downstream handling code, which no longer needs to manually probe real_stack. Maybe I should have renamed this to `maybe_get_real_stack`

ezyang replied:
Handling `stack_above_dynamo` is also shunted to this function for convenience.

ezyang commented on torch/_dynamo/exc.py:
```diff
@@ -227,32 +239,21 @@ def filter_stack(stack):
-            code.co_filename,
-            code.co_firstlineno,
-            code,
-        )
```
I have never found the bytecode printout here useful. Got rid of it. cc @williamwen42

ezyang commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -422,7 +426,7 @@ def wrapper(self: "InstructionTranslatorBase", inst: Instruction):
                 ):
                     graph_break_log.debug(
-                        "Graph break: %s from user code at %s",
+                        "Graph break: %s from user code at:\n%s",
```
This case is tested by test_graph_break_log

ezyang commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -386,6 +386,9 @@ def wrapper(self: "InstructionTranslatorBase", inst: Instruction):
             try:
+                TracingContext.set_current_loc(
+                    self.f_code.co_filename, self.lineno, self.f_code.co_name
+                )
```
This is necessary because I've removed the `self.frame_summary()` manual addition in the except handling. This handling is better as it ensures regular exceptions also get the current location setup correctly as well.

ezyang commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -684,18 +688,10 @@ def step(self):
-            real_stack = getattr(exc, "real_stack", [])
-            real_stack.append(self.frame_summary())
-            exc.real_stack = real_stack  # type: ignore[attr-defined]
-            raise
```
It's not necessary to append stack frames onto exceptions as they bubble up. Even better, it is not necessary to attach stacks either; TracingContext takes care of this.

ezyang commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -712,7 +708,10 @@ def step(self):
+        # NB: Don't push the top level frame summary; set_current_loc will
+        # take care of it.  However, DO make sure we attach real_stack to
+        # exceptions
+        return TracingContext.current_frame(None)
```
I don't think it is possible to exercise this case, but it's safest to make sure we set up the try...except handler.

ezyang committed:
Update on "Refactor stack handling to always use TracingContext to populate real stack on exception"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/_guards.py:
```diff
@@ -517,20 +523,37 @@ def clear_frame():
             yield
+        except Exception as e:
+            if not hasattr(e, "real_stack"):
+                e.real_stack = tc.extract_stack()  # type: ignore[attr-defined]
```
Attach the stack if there isn't one already. This ensures we get localized error messages for ALL exceptions, not just Unsupported.

ezyang committed:
Update on "Refactor stack handling to always use TracingContext to populate real stack on exception"


The basic gist of the PR is simple, but it's accompanied with some careful modifications and unit tests to make sure I got it right. Check inline comments for more details.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang committed:
Update on "Refactor stack handling to always use TracingContext to populate real stack on exception"


The basic gist of the PR is simple, but it's accompanied with some careful modifications and unit tests to make sure I got it right. Check inline comments for more details.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang committed:
Update on "Refactor stack handling to always use TracingContext to populate real stack on exception"


The basic gist of the PR is simple, but it's accompanied with some careful modifications and unit tests to make sure I got it right. Check inline comments for more details.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

albanD commented on torch/_guards.py:
```diff
@@ -517,20 +523,37 @@ def clear_frame():
+                yield
+            except Exception as e:
+                # Prevent real_stack from getting attached
+                if not hasattr(e, "real_stack"):
```
This is weird? Is there an extra `not` here? If so why not just set to `None` all the time?

ezyang replied:
Nope, this is correct. The invariant is that if an Exception as real_stack, we've appropriately attached a user stack and we no longer need to attach anything. Because we cannot conveniently interpose when an exception is thrown, we instead interpose everywhere we set what the user stack is set (using the context manager). However, our compiler stack does "tail calls" (when it calls into user compiler), at which point the parent exception frames would incorrectly attach an incorrect frame.

However, if, somehow, someone raised an exception with this scope that had a stack (for example, because they are restoring the user stack state appropriately as they process node by node), we should respect it. Thus, we cannot unconditionally set None.

albanD replied:
Ho so the `real_stack` field being `None` is semantically different from not existing. As we use this as a marker that a stack shouldn't be attached by a higher level handler. Ok!

voznesenskym replied:
Consider also using a placeholder/dummy object instead of None. None being morally closer to/same as not existing feels better.

ezyang replied:
But... that's exactly what it means! None means there is affirmatively no real stack (as opposed to not set, which means there may or may not be one, we don't know yet)

ezyang replied:
I went ahead and added this comment to the code

ezyang committed:
Update on "Refactor stack handling to always use TracingContext to populate real stack on exception"


The basic gist of the PR is simple, but it's accompanied with some careful modifications and unit tests to make sure I got it right. Check inline comments for more details.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang committed:
Update on "Refactor stack handling to always use TracingContext to populate real stack on exception"


The basic gist of the PR is simple, but it's accompanied with some careful modifications and unit tests to make sure I got it right. Check inline comments for more details.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on test/dynamo/test_exc.py:
```diff
@@ -0,0 +1,214 @@
+    s = re.sub(r"line \d+", "line N", s)
+    s = re.sub(file, os.path.basename(__file__), s)
+    s = re.sub(os.path.join(os.path.dirname(torch.__file__), ""), "", s)
+    s = re.sub(r"\\", "/", s)
```
yup

ezyang commented on test/dynamo/test_exc.py:
```diff
@@ -0,0 +1,214 @@
+    s = re.sub(r'  File "([^"]+)", line \d+, in (.+)\n    .+\n', repl_frame, s)
+    s = re.sub(r"line \d+", "line N", s)
+    s = re.sub(file, os.path.basename(__file__), s)
+    s = re.sub(os.path.join(os.path.dirname(torch.__file__), ""), "", s)
```
The top one gets you /path/to/pytorch/test/test_foo.py to test_foo.py. This one gets you /path/to/pytorch/ to '' (which will match files beyond the current test file.

ezyang commented on torch/_dynamo/config.py:
```diff
@@ -272,6 +272,10 @@ def is_fbcode():
 # enable/disable dynamo tracing for `torch.func` transforms
 capture_func_transforms = True
 
+# simulates what would happen if we didn't have support for BUILD_SET opcode,
```
Because then someone could add support for it and then the test would break. This only costs two lines in the prod code, worth it.

albanD reviewed (approved):
Logic is quite tricky but sounds good!

voznesenskym reviewed (approved):
We love to see it

ezyang committed:
Update on "Refactor stack handling to always use TracingContext to populate real stack on exception"


The basic gist of the PR is simple, but it's accompanied with some careful modifications and unit tests to make sure I got it right. Check inline comments for more details.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/_dynamo/convert_frame.py:
```diff
@@ -559,12 +543,45 @@ def _convert_frame(
+            if soft_fail:
+                log.info(error_msg, exc_info=True)
+            else:
+                log.warning(error_msg, exc_info=True)
```
exc_info=True here is new, I added it because @voznesenskym was complaining about missing stack traces

ezyang committed:
Update on "Refactor stack handling to always use TracingContext to populate real stack on exception"


The basic gist of the PR is simple, but it's accompanied with some careful modifications and unit tests to make sure I got it right. Check inline comments for more details.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -f "unrelated error only"
